nexus/src/app/sled.rs

```diff
@@ -237,4 +239,227 @@ impl super::Nexus {
         self.db_datastore.service_upsert(opctx, service).await?;
         Ok(())
     }
+
+    // OPTE V2P mappings
+
+    /// Ensure that the necessary v2p mappings for an instance are created
+    pub async fn create_instance_v2p_mappings(
+        &self,
+        opctx: &OpContext,
+        instance_id: Uuid,
+    ) -> Result<(), Error> {
+        // For every sled that isn't the sled this instance was allocated to, create
+        // a virtual to physical mapping for each of this instance's NICs.
+        //
+        // For the mappings to be correct, a few invariants must hold:
+        //
+        // - mappings must be set whenever an instance's sled changes (eg.
+        //   during instance creation, migration, stop + start)
+        //
+        // - an instances' sled must not change while its corresponding mappings
+        //   are being created
+        //
+        // - the same mapping creation must be broadcast to all sleds
+        //
+        // A more targeted approach would be to see what other instances share
+        // the VPC this instance is in (or more generally, what instances should
+        // have connectivity to this one), see what sleds those are allocated
+        // to, and only create V2P mappings for those sleds.
+        //
+        // There's additional work with this approach:
+        //
+        // - it means that delete calls are required as well as set calls,
+        //   meaning that now the ordering of those matters (this may also
+        //   necessitate a generation number for V2P mappings)
+        //
+        // - V2P mappings have to be bidirectional in order for both instances's
+        //   packets to make a round trip. This isn't a problem with the
+        //   broadcast approach because one of the sides will exist already, but
+        //   it is something to orchestrate with a more targeted approach.
+        //
+        // TODO-correctness OPTE currently will block instances in different
+        // VPCs from connecting to each other. If it ever stops doing this, the
+        // broadcast approach will create V2P mappings that shouldn't exist.
```

Comment:
AFAIK this is not OPTE by itself but rather just the way the firewall rules are setup.

Reply:
:+1: updated the comment in 05ee12ba